### ISSUE TITLE
SSO Authentication 404s if you have every permission

### ISSUE THREAD

**Issue by Ashterothi** [ccpgames/esi-issues#267]
The URL used in the SSO is a 404 if you use every permission available. This is likely due to URL character limit. Please add either clustered permissions, or update the scope selector to determine if your chosen scopes will fail.

**Comment by SnipersCode** [comment]
Note that the maximum query string for IIS (which I assume the webpage is using because it is a IIS 404 error page) is 2048 characters. Minimum query string length for the endpoint not including the client id and redirect uri is 2047 characters when requesting all scopes.

I recommend at least having an "all" scope

**Comment by smclenithan** [comment]
It is also worth noting that the ESI scope names are on average significantly longer than the old scopes. While an **all** scope would be useful in certain situations we still need this to be fixed as a subset of all is still likely to exceed that limit without too much effort.

Currently if you are already logged in to login.eveonline.com then this issue does not occur as the IIS login server does not receive the redirect request.

https://www.iis.net/configreference/system.webserver/security/requestfiltering/requestlimits
![image](https://cloud.githubusercontent.com/assets/1315916/22873911/312b254c-f180-11e6-9eee-275ba490787a.png)

If we could get the `maxQueryString` and `maxUrl` parameters changed to something that can accommodate all current and future scopes that would be fantastic.

**Comment by snitchashor** [comment]
Would it be possible to use POST instead of a query string? Or just have integer IDs for all the scopes?

**Comment by antihax** [comment]
While it is possible per the RFC, the library I am using for OAuth2 does not support. If POST is implemented, it would be nice to have both GET/POST supported since this is a rare case that someone would need _every_ scope.

**Comment by SnipersCode** [comment]
Corporate apps are the ones most likely and quite frequently going to need all scopes in order to grab all available information about the character and implement QoL services for members.

It should be noted that the OAuth2 Specification states that scopes needs to be a parameter in a GET request. https://tools.ietf.org/html/rfc6749#section-4.1.1

IMO, the best solutions in this case are either

1) Implement group scopes (all, all_read_only, assets, calendar, etc)
2) Change the maximum query string and url length
3) Shorten scope names

Remember that numbers 2 and 3 can cause regressions as more scopes are added.

**Comment by deadlybulb** [comment]
I deal with this now by maintaining multiple access tokens.  Not ideal, but it works well for offline apps.  The multiple authorization requests would be a pain for online apps.

**Comment by Regner** [comment]
Multiple access tokens and requesting access to things as required instead of all upfront is what I do as well.

The goal for the SSO had always been to add the ability to group scopes. Not only does that help solve this problem, but it also makes the end user experience a lot better. It cannot realistically be expected that users read through and understand all of the scopes when a large number are requested at once. Grouping them is required to help ensure users understand what they are granting applications access to.

**Comment by lelandstarr** [comment]
While the RFC does clearly state that scope needs to be a request parameter, there is no verbiage stating that it has to be a query parameter.  Unless I am missing something, there is no reason that this couldn't also be accepted as a "application/x-www-form-urlencoded" form parameter written to the get request and be fully compliant.   I admit for an actual application you shouldn't need every single scope, however some of us (most of us) are building apps for ourselves, or playing around, and having all scopes available, at least until we have a finished product is quite a bit more convenient.

**Comment by Robbilie** [comment]
this is still a SSO and not ESI issue…

**Comment by SnipersCode** [comment]
While I agree this is primarily an SSO issue, the current issue is that the scopes for esi are much longer than the ones for crest.

@lelandstarr : According to the wording of the RFC that I had linked to above,

> The client constructs the request URI by adding the following
   parameters to the **query component** of the authorization endpoint URI
   using the "application/x-www-form-urlencoded" format...

**(emphasis mine)**
While some packages have support for using the scope as a form parameter, not all do (as stated by antihax). The fact that the RFC states that it should be a "query component" implies, at least to me, that it must be a query parameter in the URI.

**Comment by tyler-sommer** [comment]
I ran into this problem in the last few days, glad to see it's already known.

@SnipersCode/@lelandstarr: the OAuth authorize endpoint itself does need to receive the values in the query string, but it seems (as of today) that its not this endpoint failing. It's when the user is redirected to the regular https://login.eveonline.com/Account/LogOn page to enter their EVE user/pass, the "RedirectUrl" parameter (containing the full original OAuth Path+query params url encoded) causes the site to throw the 404.

Since the the OAuth authorize endpoint is on the same domain (login.eveonline.com), it may be possible to simply put the RedirectURL in the browser's cookies. But still, this would require changes to the login.eveonline.com site, so that might be out of the question.

I'd love to see this issue fixed, but the mentioned workarounds seem reasonable enough. You can also get around the issue if you ensure your browser is logged into EVE before attempting to authenticate your application. Not the best, but works well enough for development and testing.

**Comment by OrangeDrangon** [comment]
It would be nice if it was fixed via grouping so that users could know what they are agreeing to better and this problem could be avoided.

**Comment by evanova** [comment]
Can we have a hope of a resolution soon?

Being in the Android world, I cannot fiddle with the browser, I cannot possibly ask the user to log-in multiple times and maintain multiple tokens, and I am blocked in moving features from the XML API because the list of scopes I require is now getting too long.

Please?

**Comment by EvilKot** [comment]
Just joined you guys and
literally unplayable

**Comment by mglinski** [comment]
Who did you join?

I have no quarrel with the rest of your comment.

**Comment by mglinski** [comment]
For everyone else waiting on this to be fixed, [CCP have said that they are in the process of "unfucking ISS"](https://github.com/ccpgames/esi-issues/issues/483#issuecomment-314463867) which I greatly commend them for and that the change is currently being evaluated on their staging env with no promised date for deployment to production. SoonTM

**Comment by EvilKot** [comment]
@mglinski 
> Who did you join?

Well, i meant i'm another guy who faced the same problem.
Also, thank you for providing link for the CCP response.

**Comment by Makeshift** [comment]
It might be worth noting that this happens when the user isn't logged in already due to the redirect to the login page doubling the length of the URL. https://gist.github.com/Makeshift/edb96d933225013b508d08bf7ec11562

Note newline is where the URL doublues. If the user is already logged in, no redirect happens and it works as intended.

**Comment by porowns** [comment]
I have also run into this issue.

My url has a character length of 2999, for users that are **already logged into EVE,** it works. However, if you haven't logged into EVE, users **receive a 404.**

**Comment by tsuthers-ccpgames** [comment]
This has been partially fixed. The problem with the login screen doubling every scope has been fixed, which has functionally raised the character limit to the point where you can fit all the ESI scopes in one request. That should unblock anyone who has pending work blocked by the inability to make login requests with all ESI scopes.

If you add literally every third party scope in existence to your application it is still possible to hit a character limit, but there is a small amount of headroom that may allow some crest scopes to be mixed in with the ESI ones (ballpark 200 characters, I think).

However, we're supposed to be at crest parity, so if you have use cases for mixing ESI and crest scopes I'm super-keen to know what they are, so we can move them to ESI and let you remove those scopes.

Meantime, I'm leaving this ticket open because we still need to raise the URL limit higher for a longer term fix.

**Comment by neophenix** [comment]
You might be at CREST parity as far as "you can do the same general things in ESI as CREST", however the data returned is not 1 to 1.  ESI mostly just returns IDs whereas CREST would give you the characterID and name.

This means that anything coded against CREST, if it is still in use would need some migration period to ESI, which means requesting both scopes.  And therein lies the problem with needing to request a large number of scopes and the 404s

**Comment by a-tal** [comment]
url limits have been raised. going forward please use https://github.com/ccpgames/sso-issues/issues for SSO related issues
